hudi-common/src/test/java/org/apache/hudi/common/table/read/SchemaHandlerTestBase.java

```diff
@@ -0,0 +1,382 @@
+    } else {
+      when(hoodieTableConfig.getPreCombineField()).thenReturn(null);
+    }
+    if (mergeMode == CUSTOM) {
+      when(hoodieTableConfig.getRecordMergeStrategyId()).thenReturn("asdf");
+      // NOTE: in this test custom doesn't have any meta cols because it is more interesting of a test case
+      when(hoodieTableConfig.getTableVersion()).thenReturn(HoodieTableVersion.EIGHT);
+    }
+  }
+
+  private static HoodieRecordMerger mockRecordMerger(boolean isProjectionCompatible, String[] mandatoryFields) throws IOException {
+    HoodieRecordMerger merger = mock(HoodieRecordMerger.class);
+    when(merger.isProjectionCompatible()).thenReturn(isProjectionCompatible);
+    when(merger.merge(any(), any(), any(), any(), any())).thenReturn(null);
+    when(merger.getMandatoryFieldsForMerging(any(), any(), any())).thenReturn(mandatoryFields);
+    return merger;
+  }
+
+  static HoodieReaderContext<String> createReaderContext(HoodieTableConfig hoodieTableConfig, boolean supportsParquetRowIndex, boolean hasLogFiles,
+                                                         boolean hasBootstrapBaseFile, boolean mergeUseRecordPosition, HoodieRecordMerger merger) {
+    HoodieReaderContext<String> readerContext = new MockReaderContext(hoodieTableConfig, supportsParquetRowIndex);
+    readerContext.setHasLogFiles(hasLogFiles);
+    readerContext.setHasBootstrapBaseFile(hasBootstrapBaseFile);
+    readerContext.setShouldMergeUseRecordPosition(mergeUseRecordPosition);
+    readerContext.setRecordMerger(Option.ofNullable(merger));
+    return readerContext;
+  }
+
+  static FileGroupReaderSchemaHandler createSchemaHandler(HoodieReaderContext<String> readerContext, Schema dataSchema,
+                                                          Schema requestedSchema, HoodieTableConfig hoodieTableConfig,
+                                                          boolean supportsParquetRowIndex) {
+    return supportsParquetRowIndex
+        ? new ParquetRowIndexBasedSchemaHandler(readerContext, dataSchema, requestedSchema,
+        Option.empty(), hoodieTableConfig, new TypedProperties())
+        : new FileGroupReaderSchemaHandler(readerContext, dataSchema, requestedSchema,
+        Option.empty(), hoodieTableConfig, new TypedProperties());
+  }
+
+  static Schema generateProjectionSchema(String... fields) {
+    return HoodieAvroUtils.generateProjectionSchema(DATA_SCHEMA, Arrays.asList(fields));
+  }
+
+  private static Schema generateProjectionSchema(boolean hasBuiltInDelete, String... fields) {
+    List<String> fieldList = Arrays.asList(fields);
+    if (hasBuiltInDelete) {
+      fieldList = new ArrayList<>(fieldList);
+      fieldList.add("_hoodie_is_deleted");
+    }
+    return HoodieAvroUtils.generateProjectionSchema(DATA_SCHEMA, fieldList);
+  }
+
+  Schema.Field getField(String fieldName) {
+    return DATA_SCHEMA.getField(fieldName);
+  }
+
+  static class MockReaderContext extends HoodieReaderContext<String> {
```

Comment:
Is it possible to swap this for an actual mock? if not, I typically call a class like this "Stubbed" instead of a mock to avoid confusion.

Reply:
Addressed. It will not be possible to mock since the class is modified by the handler.